plonky3/src/circuit_builder.rs

```diff
@@ -283,42 +255,24 @@ impl<'a, T: FieldElement, AB: AirBuilderWithPublicValues<F = Val> + PairBuilder>
         let main = builder.main();
         let fixed = builder.preprocessed();
         let pi = builder.public_values();
-        let publics = self.get_publics();
-        assert_eq!(publics.len(), pi.len());
+        assert_eq!(self.publics.len(), pi.len());
+
+        let local = main.row_slice(0);
 
         // public constraints
         let pi_moved = pi.to_vec();
-        let (local, next) = (main.row_slice(0), main.row_slice(1));
-
-        let public_offset = self.analyzed.commitment_count();
-
-        publics.iter().zip(pi_moved).enumerate().for_each(
-            |(index, ((_, col_id, row_id), public_value))| {
-                //set decr for each public to be row_id in the first row and decrement by 1 each row
-                let (decr, inv_decr, s, decr_next) = (
-                    local[public_offset + 3 * index],
-                    local[public_offset + 3 * index + 1],
-                    local[public_offset + 3 * index + 2],
-                    next[public_offset + 3 * index],
-                );
-
-                let mut when_first_row = builder.when_first_row();
-                when_first_row.assert_eq(
-                    decr,
-                    cast_to_goldilocks(GoldilocksField::from(*row_id as u32)),
-                );
-
-                let mut when_transition = builder.when_transition();
-                when_transition.assert_eq(decr, decr_next + AB::Expr::one());
-
-                // is_zero logic-- s(row) is 1 if decr(row) is 0 and 0 otherwise
-                builder.assert_bool(s); //constraining s to 1 or 0
-                builder.assert_eq(s, AB::Expr::one() - inv_decr * decr);
-                builder.assert_zero(s * decr); //constraining is_zero
+        let pi_deref = pi_moved.clone();
+        let fixed_local = fixed.row_slice(0);
+        let public_offset = self.analyzed.constant_count();
+
+        self.publics.iter().zip(pi_moved).enumerate().for_each(
+            |(index, ((_, col_id, _), public_value))| {
+                let selector = fixed_local[public_offset + index];
+                builder.assert_bool(selector);
```

Comment:
If the selector is a fixed column we do not need to constrain it inside the circuit